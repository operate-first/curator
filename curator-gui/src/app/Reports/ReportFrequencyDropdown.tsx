import React, { EventHandler } from 'react';

type myProps = {
  setReportFrequency: Function;
  ReportFrequency: string;
}

type myState = {
}

/*
Dropdown component for the Reports page to specify one of three report frequencies:
Daily, Weekly, and Monthly.
*/
class ReportFrequencyDropdown extends React.Component<myProps,myState>{
  constructor(myProps) {
    super(myProps);
    this.state = {
    }
  }

  ReportFrequency=(event: React.FormEvent<HTMLSelectElement>)=>{
    this.props.setReportFrequency(event.currentTarget.value)
  }

  createDropDowns(key: string, reportFrequencies: Array<string>, onChange: Function, def: string ) {
    return(
      <React.Fragment>
        <label>Select Report Frequency</label>
      <select key={key} onChange={e=>onChange(e)} value={def} aria-label="Report Frequency">
         
        {reportFrequencies.map((i)=>{ 
          return(<option key={key+i} value={i}>{i}</option>)
        } )}
      </select>

      </React.Fragment>      
    )  
  }

  render() {
    const reportTypes: Array<string> = ['day', 'week', 'month']
    return (
      
      <div>
        {this.createDropDowns("Report Frequency",reportTypes,this.ReportFrequency,this.props.ReportFrequency)}
      </div>
    );
  }
}

export default ReportFrequencyDropdown
