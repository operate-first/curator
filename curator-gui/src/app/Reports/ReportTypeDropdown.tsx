import React, { EventHandler } from 'react';

type myProps = {
  setReportType: Function;
  ReportType: string;
}

type myState = {
}

/*
Dropdown component for the Reports page to specify one of three report types:
Standard, Analytical, and Custom. 
*/
class ReportTypeDropdown extends React.Component<myProps,myState>{
  constructor(myProps) {
    super(myProps);
    this.state = {
    }
  }

  ReportType=(event: React.FormEvent<HTMLSelectElement>)=>{
    this.props.setReportType(event.currentTarget.value)
  }

  createDropDowns(key: string, reportTypes: Array<string>, onChange: Function, def: string ) {
    return(
      <React.Fragment>
        <label>Select Report Type</label>
      <select key={key} onChange={e=>onChange(e)} value={def} aria-label="Report Type">
         
        {reportTypes.map((i)=>{ 
          return(<option key={key+i} value={i}>{i}</option>)
        } )}
      </select>

      </React.Fragment>
      
    )    
  }

  render() {
    const reportTypes: Array<string> = ['standard', 'analytics', 'custom']
    return (   
      <div>
        {this.createDropDowns("Report Type",reportTypes,this.ReportType,this.props.ReportType)}
      </div>
    );
  }
}

export default ReportTypeDropdown
