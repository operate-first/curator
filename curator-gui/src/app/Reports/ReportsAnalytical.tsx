import '@patternfly/react-core/dist/styles/base.css';
import React, { Component, ObjectHTMLAttributes } from "react";
import ReactApexChart from "react-apexcharts";

type myProps = {
  indexNum: number,
  columnTitle: object;
  cpuUsage: object;
  networkUsage: object;
  memoryUsage: object;
  startDate: Date;
  reportFrequency: string;
  namespace: string;
};

type myState = {
  options: any;
  series: any;
};

// Represents a line graph for one namespace with lines detailing 
// CPU, Network, and Memory usage. 
class ReportsAnalytical extends React.Component<myProps, myState> {
  constructor(props) {
    super(props);

    this.state = {
      series: [{
        name: 'CPU Usage',
        type: 'line',
        data: this.props.cpuUsage
      }, {
        name: 'Network Usage',
        type: 'line',
        data: this.props.networkUsage
      }, {
        name: 'Memory Usage',
        type: 'line',
        data: this.props.memoryUsage
      }],
      options: {
        chart: {
          height: 350,
          type: 'line',
          stacked: false
        },
        dataLabels: {
          enabled: false
        },
        stroke: {
          width: [1, 1, 4]
        },
        title: {
          text: 'Namespace: ' + this.props.namespace + ' Start Date: ' + this.props.startDate + ' '
            + this.props.reportFrequency + ': ' + this.props.indexNum,
          align: 'left',
          offsetX: 110
        },
        xaxis: {
          categories: this.generateXScale(this.props.startDate, this.props.reportFrequency),
        },
        yaxis: [
          { // CPU Usage 
            seriesName: 'CPU Usage',
            axisTicks: {
              show: true,
            },
            axisBorder: {
              show: true,
              color: '#008FFB'
            },
            labels: {
              style: {
                colors: '#008FFB',
              }
            },
            title: {
              text: "CPU Usage in Seconds",
              style: {
                color: '#008FFB',
              }
            },
            // tooltip: {
            //   enabled: true
            // }
          },
          {
            seriesName: 'Network Usage',
            opposite: true,
            axisTicks: {
              show: true,
            },
            axisBorder: {
              show: true,
              color: '#00E396'
            },
            labels: {
              style: {
                colors: '#00E396',
              }
            },
            title: {
              text: "Network Usage in Megabits per Second",
              style: {
                color: '#00E396',
              }
            },
          },
          {
            seriesName: 'Memory Usage',
            opposite: true,
            axisTicks: {
              show: true,
            },
            axisBorder: {
              show: true,
              color: '#FEB019'
            },
            labels: {
              style: {
                colors: '#FEB019',
              },
            },
            title: {
              text: "Memory Usage in Gigabytes",
              style: {
                color: '#FEB019',
              }
            }
          },
        ],
        tooltip: {
          fixed: {
            enabled: true,
            position: 'topLeft', // topRight, topLeft, bottomRight, bottomLeft
            offsetY: 30,
            offsetX: 60
          },
        },
        legend: {
          horizontalAlign: 'left',
          offsetX: 40
        }
      },
    };
  }

  // returns an array which represents a time scale
  // forthe x axis.
  generateXScale(startDate: Date, frequency: string) {
    const xScale: Array<string> = [];

    switch (frequency) {
      case "monthly":
        for (var i = 0; i < 4; i++) {
          //TODO: generate a scale of 4 or so weeks
        }
        break;
      case "weekly":
        for (var i = 0; i < 7; i++) {
          xScale.push(new Date(startDate.getTime() + i * 24 * 60 * 60 * 1000).toDateString());
        }
        break;
      default: //daily as default 
        for (var i = 0; i < 4; i++) {
          //TODO: generate a scale of 24 hours
        }
    }
    return xScale;
  }

  render() {
    return (
      <div className="app">
        <div className="row">
          <div className="mixed-chart">
            <ReactApexChart
              options={this.state.options}
              series={this.state.series}
              type="line"
              width={1000}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default ReportsAnalytical;