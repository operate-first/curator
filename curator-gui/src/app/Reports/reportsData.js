const reportsData =[
    {"namespace":"hello-world","activationTime":63},
    {"namespace":"mywebapp","activationTime":63},
    {"namespace":"openshift-apiserver","activationTime":63},
    {"namespace":"openshift-apiserver-operator","activationTime":63},
    {"namespace":"openshift-authentication","activationTime":63},
    {"namespace":"openshift-authentication-operator","activationTime":63},
    {"namespace":"openshift-cluster-machine-approver","activationTime":63},
    {"namespace":"openshift-cluster-node-tuning-operator","activationTime":63},
    {"namespace":"openshift-cluster-samples-operator","activationTime":63},
    {"namespace":"openshift-cluster-storage-operator","activationTime":63},
    {"namespace":"openshift-console","activationTime":63},
    {"namespace":"openshift-console-operator","activationTime":63},
    {"namespace":"openshift-controller-manager","activationTime":63},
    {"namespace":"openshift-controller-manager-operator","activationTime":63},
    {"namespace":"openshift-dns","activationTime":63},
    {"namespace":"openshift-dns-operator","activationTime":63},
    {"namespace":"openshift-etcd","activationTime":63},
    {"namespace":"openshift-image-registry","activationTime":63},
    {"namespace":"openshift-ingress","activationTime":63},
    {"namespace":"openshift-ingress-operator","activationTime":63},
    {"namespace":"openshift-kube-apiserver","activationTime":63},
    {"namespace":"openshift-kube-apiserver-operator","activationTime":63},
    {"namespace":"openshift-kube-controller-manager","activationTime":63},
    {"namespace":"openshift-kube-controller-manager-operator","activationTime":63},
    {"namespace":"openshift-kube-scheduler","activationTime":63},
    {"namespace":"openshift-kube-scheduler-operator","activationTime":63},
    {"namespace":"openshift-machine-config-operator","activationTime":63},
    {"namespace":"openshift-marketplace","activationTime":63},
    {"namespace":"openshift-metering","activationTime":63},
    {"namespace":"openshift-monitoring","activationTime":63},
    {"namespace":"openshift-multus","activationTime":63},
    {"namespace":"openshift-network-operator","activationTime":63},
    {"namespace":"openshift-operator-lifecycle-manager","activationTime":63},
    {"namespace":"openshift-sdn","activationTime":63},
    {"namespace":"openshift-service-ca","activationTime":63},
    {"namespace":"openshift-service-ca-operator","activationTime":63},
    {"namespace":"openshift-service-catalog-apiserver-operator","activationTime":63},
    {"namespace":"openshift-service-catalog-controller-manager-operator","activationTime":63},
    {"namespace":"sample","activationTime":63},
    {"namespace":"hello-world-demo","activationTime":40}]