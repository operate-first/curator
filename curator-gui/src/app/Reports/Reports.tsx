import * as React from 'react';
import {
    PageSection,
    Title,
    Stack,
    StackItem,
    Button
} from '@patternfly/react-core';
import { ReportsDataFilterForm } from './ReportsDataFilterForm';

type myProps = {};
type myState = {
    startHrs: number;
    endHrs: number;
    startDate: Date;
    endDate: Date;
};

/*
Component which allows users to generate and view metering reports.
*/
class Reports extends React.Component<myProps, myState> {
    constructor(myProps) {
        super(myProps);
        this.state = {
            startHrs: 0,
            endHrs: 0,
            startDate: new Date(),
            endDate: new Date(),
        }
    }

    render() {
        return (
            <PageSection>
                <Title headingLevel="h1" size="lg">
                    Resource Usage
                </Title>
                <Stack>
                    <StackItem>
                        <ReportsDataFilterForm />
                    </StackItem>
                </Stack>
            </PageSection>
        );
    }
}

export { Reports };
