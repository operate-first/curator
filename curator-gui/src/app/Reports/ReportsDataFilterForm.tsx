import React from 'react';
import { Grid, GridItem, Form, ActionGroup, Alert } from '@patternfly/react-core';
import axios from 'axios';
import { SimpleInputGroups } from '@app/DateComponent/DateComponent';
import { Button } from '@patternfly/react-core';
import ReportsList from './ReportsList';
import CsvDownload from 'react-json-to-csv';
import ReportTypeDropdown from './ReportTypeDropdown';
import ReportFrequencyDropdown from './ReportFrequencyDropdown';
import ReportsAnalytical from './ReportsAnalytical';

type myProps = {};
type myState = {
    startDate: Date;
    conditionalRender: number;
    changingDate: boolean;
    api: string;
    clusterData: Array<dataObject> | null;
    err: string | null;
    isLoaded: boolean;
    reportType: string;
    changingReportType: boolean;
    reportFrequency: string;
    changingReportFrequency: boolean;
};
export type dataObject = {
    namespace: string;
    podUsageCpuCoreSeconds: number;

};

const convertDateToUTC = (date: Date) => {
    return Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), date.getUTCHours(), 0, 0, 0);
};

/*
Form for users to specify the start date, report frequency(daily, weekly, monthly), and report type
(standard, analytical, custom) of a metering report. Also offers an option to export metering data from the
backend as csv.
*/
class ReportsDataFilterForm extends React.Component<myProps, myState> {
    constructor(myProps) {
        super(myProps);

        this.state = {
            startDate: new Date(),
            conditionalRender: 0,
            changingDate: true,
            api: 'http://localhost:8000/namespace_cpu_request',
            clusterData: null,
            err: null,
            isLoaded: false,
            reportType: 'standard',
            changingReportType: false,
            reportFrequency: 'day',
            changingReportFrequency: false
        };

        this.callAPI(false);
    }


    shouldComponentUpdate(nextProps, nextState) {
        return JSON.stringify(this.state) !== JSON.stringify(nextState);
    }


    callAPI(onSubmit) {
        const startDate = new Date(convertDateToUTC(this.state.startDate)).toISOString().split('T', 1)[0];

        let apiUrl = this.state.api;
        if (onSubmit) {
            apiUrl = apiUrl + '/' + this.state.reportType + '?start=' + startDate + '&frequency=' + this.state.reportFrequency;
        }

        axios
            .get(apiUrl)
            .then(res => {
                this.setState({ ...this.state, isLoaded: true, clusterData: res.data });
            })
            .catch(err => {
                this.setState({ ...this.state, isLoaded: false, err: err });
            });
    }

    changeToggle() {
        this.callAPI(true);
    }

    setDate = (date: Date) => {
        date = new Date(date);
        date.setDate(date.getDate() + 1);
        this.setState({ ...this.state, changingDate: true, startDate: new Date(date) });
    };

    setReportType = (aType: string) => {
        this.setState({ ...this.state, changingReportType: true, reportType: aType });
    };

    setReportFrequency = (aFrequency: string) => {
        this.setState({ ...this.state, changingReportFrequency: true, reportFrequency: aFrequency });
    };

    renderReport = () => {
        switch (this.state.reportType) {
            case 'standard':
                return (this.renderStandard());
            case 'analytics':
                return (this.renderAnalytical());
            case '':
                return (this.renderReportingError('Unable to generate report. Report Type not specified.'));
            default:
                return (this.renderReportingError('Unable to generate report. Report Type not yet supported.'));
        }
    }

    // Renders a error banner which displays a provided string value
    renderReportingError(message: string) {
        return (
            <React.Fragment>
                <Alert variant="danger" isInline title={message} />
            </React.Fragment>
        )
    }

    // Generates a standard report, a table with columns for namespace, cpu usage, network usage,
    // and memory usage.
    renderStandard = () => {
        const columnTitle = {
            namespace: 'namespace',
            podUsageCpuCoreSeconds: 'Pod CPU Usage in Seconds',

        };

        const tableData: Array<dataObject> = [];

        if (this.state.clusterData !== null) {
            this.state.clusterData['reports'].forEach(clusterInfo => {
                tableData.push({
                    namespace: clusterInfo['namespace'],
                    // Note: The line below should be `podUsageCpuCoreSeconds: clusterInfo['pod_usage_cpu_core_seconds']`
                    // The mock API call only supports 'pod_request_cpu_core_seconds' but
                    // I want to ensure that the value gets rendered.
                    podUsageCpuCoreSeconds: clusterInfo['pod_request_cpu_core_seconds'],
                })
            })
        }

        return (
            <div>
                {this.state.clusterData !== null && (
                    <ReportsList
                        key={'ReportsList'}
                        startDate={this.state.startDate}
                        columnTitle={columnTitle}
                        tableData={tableData}
                    />
                )}
            </div>
        );
    };

    // Generates an analytical report: line graphs for each namespace with
    // lines denoting CPU, network, and memory usage.
    renderAnalytical = () => {
        const columnTitle = {};
        let idx = 0;
        return (
            <div>
                {this.state.clusterData !== null && (this.state.clusterData['reports']
                    .map((value, idx) => {
                        idx++;
                        return (
                            <ReportsAnalytical
                                key={idx}
                                indexNum={idx}
                                columnTitle={columnTitle}
                                cpuUsage={value['cpu_usage']}
                                networkUsage={value['network_usage']}
                                memoryUsage={value['memory_usage']}
                                namespace={value['namespace']}
                                startDate={this.state.startDate}
                                reportFrequency={this.state.reportFrequency} />)
                    }))}
            </div>
        )
    }

    render() {
        return (
            <React.Fragment>
                <Form>
                    <div>Select a date range to view available Daily, Weekly, and Monthly reports.</div>
                    <Grid>
                        <GridItem span={8}>
                            <SimpleInputGroups changeDate={this.setDate} dateType="Date" key="Date" />
                        </GridItem>
                    </Grid>
                    <Grid>
                        <GridItem span={6}>
                            <CsvDownload data={this.state.clusterData}>Download as CSV</CsvDownload>
                        </GridItem>
                    </Grid>
                    <Grid>
                        <GridItem span={8}>
                            <ReportTypeDropdown setReportType={this.setReportType} ReportType={this.state.reportType} />
                        </GridItem>
                    </Grid>
                    <Grid>
                        <GridItem span={8}>
                            <ReportFrequencyDropdown setReportFrequency={this.setReportFrequency} ReportFrequency={this.state.reportFrequency} />
                        </GridItem>
                    </Grid>
                    <Grid>
                        <ActionGroup>
                            <GridItem span={6}>
                                <Button onClick={() => this.changeToggle()}>Submit</Button>
                            </GridItem>
                        </ActionGroup>
                    </Grid>
                    <Grid>
                        <GridItem span={10} rowSpan={8}>
                            {this.state.isLoaded && this.renderReport()}
                            {!this.state.isLoaded && this.state.err !== null && <div>{this.state.err.toString()}</div>}
                        </GridItem>
                    </Grid>
                </Form>
            </React.Fragment>
        );
    }
}

export { ReportsDataFilterForm };
