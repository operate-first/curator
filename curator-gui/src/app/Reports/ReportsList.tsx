import React from 'react';
import '@patternfly/react-core/dist/styles/base.css';
import { Table, TableHeader, TableBody, TableVariant, BodyCell } from '@patternfly/react-table';
import { Link } from 'react-router-dom';
import { dataObject } from './ReportsDataFilterForm';

type myProps = {
  columnTitle: object;
  tableData: Array<dataObject>;
  startDate: Date;
};

type myState = {
  columns: Array<object | string>;
  rows: Array<row>;
};

type row = {
  cells: cells;
};

type cells = Array<JSX.Element | number | string>;

/*
The table/list which displays metering report data for the Standard Report Type.
Currently displays columns for namespace, cpu usage, network usage, and memory usage.
*/
class ReportsList extends React.Component<myProps, myState> {
  constructor(myProps) {
    super(myProps);

    const rowData: Array<row> = [];
    myProps.tableData.forEach(dataRow => {
      rowData.push({
        cells: [
          <BodyCell key={`/projectlist/${dataRow['namespace']}`}>
            <Link to={`/projectlist/${dataRow['namespace']}`} key={`/projectlist/${dataRow['namespace']}`}>
              {dataRow['namespace']}
            </Link>
          </BodyCell>,
          dataRow['podUsageCpuCoreSeconds'],

        ]
      });
    })

    this.state = {
      columns: [
        myProps.columnTitle['namespace'],
        myProps.columnTitle['podUsageCpuCoreSeconds']
      ],
      rows: rowData
    };
  }

  shouldComponentUpdate(nextProps: myProps,nextState: myState){
    console.log(nextProps.tableData);
    console.log(JSON.stringify(nextProps)!== JSON.stringify(this.props))
    return JSON.stringify(nextProps)!== JSON.stringify(this.props)
  }

  UNSAFE_componentWillReceiveProps(nextProps: myProps) {
      const rowData: Array<row> = [];

      nextProps.tableData.forEach(dataRow => {
        rowData.push({
          cells: [
            <BodyCell key={`/projectlist/${dataRow['namespace']}`}>
              <Link to={`/projectlist/${dataRow['namespace']}`} key={`/projectlist/${dataRow['namespace']}`}>
              {dataRow['namespace']}
              </Link>
            </BodyCell>,
            dataRow['podUsageCpuCoreSeconds']
          ]
        });
      });
      this.setState({ ...this.state, rows: rowData });
  }

  render() {
    return (
      <div>
        <Table
          key={'dataTable'}
          aria-label="Compact Table"
          variant={TableVariant.compact}
          cells={this.state.columns}
          rows={this.state.rows}
          caption="List of reports"
        >
          <TableHeader />
          <TableBody />
        </Table>
      </div>
    );
  }
}
export default ReportsList;
