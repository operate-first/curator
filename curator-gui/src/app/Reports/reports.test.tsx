import React from 'react';
import { shallow } from 'enzyme';
import { Reports } from './Reports';

describe('Reports tests', () => {

    // Reports currently does not use state, so further tests are not needed
    test('should render default Reports component', () => {
      const view = shallow(<Reports />);
      expect(view).toMatchSnapshot();
    })
  
  })