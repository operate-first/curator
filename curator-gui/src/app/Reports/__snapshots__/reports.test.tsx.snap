// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reports tests should render default Reports component 1`] = `
<Component>
  <Component
    headingLevel="h1"
    size="lg"
  >
    Reports
  </Component>
  <Component>
    <Component>
      <ReportsDataFilterForm />
    </Component>
  </Component>
</Component>
`;
